doc/contribute/coding_guidelines/index.rst

```diff
@@ -1297,6 +1297,85 @@ macros. Given that Zephyr uses a fork of the corresponding upstream for each
 module, it is always possible to patch the macro implementation in each module
 to avoid collisions.
 
+.. _coding_guideline_libc_usage_restrictions_in_zephyr_kernel:
+
+Rule A.4: C Standard Library Usage Restrictions in Zephyr Kernel
+================================================================
+
+Severity
+--------
+
+Required
+
+Description
+-----------
+
+Zephyr kernel shall not use the C standard library functions and macros that
+are not available in the :ref:`minimal libc <c_library_minimal>`.
+
+The "Zephyr kernel" in this context consists of the following components:
+
+* Kernel (:file:`kernel`)
+* OS Library (:file:`lib/os`)
+
+The full list of the C standard library functions that are available as part of
+the minimal libc can be found :ref:`here <c_library_minimal_functions>`.
+
+Rationale
+---------
+
+Zephyr kernel must be able to build with the
+:ref:`minimal libc <c_library_minimal>`, a limited C standard library
+implementation that is part of the Zephyr RTOS and maintained by the Zephyr
+Project, to allow self-contained testing and verification of the kernel and
+core OS services.
+
+In order to ensure that the Zephyr kernel can build with the minimal libc, it
+is necessary to restrict the use of the C standard library functions and macros
+in the Zephyr kernel to the functions and macros that are available as part of
+the minimal libc.
+
+Rule A.5: C Standard Library Usage Restrictions in Zephyr Codebase
+==================================================================
+
+Severity
+--------
+
+Required
+
+Description
+-----------
+
+Zephyr codebase shall not use the functions and macros that are not defined by
+the ISO C standard, unless exempted by this rule.
+
+The "Zephyr codebase" in this context refers to all source code files committed
```

Comment:
Yup, seems like an exception needs to be added which allows use of functions from the POSIX extensions to the C library as defined by IEEE Std 1003.1-2017.

Reply:
> Yup, seems like an exception needs to be added which allows use of functions from the POSIX extensions to the C library as defined by IEEE Std 1003.1-2017.

And presumably, those should be on an "approved" list as well.

Reply:
The scope of this is the functions and macros **from the libc** as implied by the name of this rule: "Rule A.5: C Standard Library Usage Restrictions in Zephyr Codebase."

So basically, "Zephyr codebase shall not use the **C standard library-provided** functions and macros that are not defined by the ISO C standard, unless exempted by this rule."

Since the Zephyr POSIX library functions are provided by the Zephyr POSIX library (i.e. not libc/C standard library), that is outside the scope of this rule and therefore not disallowed by this rule.

I will try to reword this to make that more clear.

Reply:
Perhaps we need a separate specification for which POSIX functions are supported by the Zephyr POSIX extension?

Reply:
> Perhaps we need a separate specification for which POSIX functions are supported by the Zephyr POSIX extension?

https://docs.zephyrproject.org/latest/services/portability/posix.html#units-of-functionality might be what you are looking for.

Reply:
yup, sorry for not looking around.